Allow a file to be moved cross-device.

miku opened the pull request:
<!--- This template is optional. Please use it as a starting point to help guide PRs -->

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes -->

Implementation of `luigi.File` would now allow movement of a file across filesystems.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

If you use `luigi.File(src).move(target)` when source and target are on different filesystems, you get an error:

```
OSError: [Errno 18] Invalid cross-device link
```

Also, this would align implementation of `File.move()` with `File.copy()` which already uses `shutil`.

## Have you tested this? If so, how?
<!--- Valid responses are "I have included unit tests." or --> 
<!--- "I ran my jobs with this code and it works for me." -->

Semantic stays the same. Existing tests apply.

> If the destination is on the current filesystem, then os.rename() is used. - [shutil.move](https://docs.python.org/3/library/shutil.html#shutil.move)

Important tradeoff: If you move a file across filesystems, you lose atomicity of `os.rename`.

erikbern commented:
lgtm

miku commented:
Thanks for the review, @Tarrasch, I tried to address your points. As for the docs, I didn't find a perfect place yet, so added a short section to the tasks docs.

erikbern commented:
> Mention the atomicity things in the docs? Atomicity of renames are a very important thing in luigi. If it can't be maintained for some functions, the docs should say under what condition.

pretty sure cross device file moving will never be atomic

miku commented:
@erikbern, closest thing to atomic would probably be: copy file to a temp location on the target device, rename it to destination, then unlink the source? But I wanted to keep the code short.

Luigi allows non-atomicity, e.g. in tasks with multiple outputs:

> If multiple outputs are returned, atomicity will be lost ... -- [Task.output](https://luigi.readthedocs.io/en/stable/tasks.html#task-output)

So I thought it would be ok to drop atomicity if this special case, if the user chooses to (chooses to work with two file systems).

Tarrasch commented on test/file_test.py:
```diff
@@ -130,6 +130,24 @@ def test_move(self):
+        t = LocalTarget(self.path)
+        f = t.open('w')
+        f.write('test_data')
+        f.close()
```
Hmm, can't you do the much more elegant context-manager thingie? `with`-statements.

Tarrasch commented on doc/tasks.rst:
```diff
@@ -307,3 +307,19 @@ Luigi also does some metaclass logic so that
+
+.. code:: python
+
+    luigi.LocalTarget(path=src).move(dst)
```
Wouldn't it be much more easy to put this in the docs for `LocalTarget.move()` docs?? You don't even need a code example. Just a warning saying to be aware of the risks when doing this across filesystems's.

miku replied:
Yes, you're right. I addressed this the [next PR](https://github.com/spotify/luigi/pull/1877/commits/93f064d3bc0a7602b65a6af532a7f0f20f2c7dc0).

erikbern commented:
> @erikbern, closest thing to atomic would probably be: copy file to a temp location on the target device, rename it to destination, then unlink the source? But I wanted to keep the code short.

great idea, let's do this

miku commented:
> great idea, let's do this

Thanks for the feedback. I implemented an almost-atomic cross-device move operation and updated the docs. (I'll be AFK for ten days, can only look into this when I'm back).

Tarrasch commented:
Does this work for you in production as well? If so, let's merge. :)

erikbern commented:
lgtm – nice!

miku commented:
> Does this work for you in production as well?

@Tarrasch, yes it does.

Background: We have `$TEMPDIR` and data on different partitions plus we sometimes do writes to a tempfile first, then *manually* move the file into place with a 

    luigi.LocalTarget(sometempfile).move(self.output().path)

miku commented:
Thanks!
